Enhance created lwc

AmigoBongero opened the pull request:
- Dev: my Accounts and Recently Viewed Accounts tabs 
- Dev: apex controller for accounts component 

Time spent: 5 hours

AmigoBongero committed:
# Wrote MyAccounts tab

- dev: MyAccounts tab functuinality
- dev: myAccounts Apex Controller
- fix: gitignore

AmigoBongero committed:
# Developed recently viewed accounts tab

- Dev: recently viewed accounts tab
- Dev: apex method for recently viewed accounts tab

IJWMG commented on DevJourney/.gitignore:
```diff
@@ -52,3 +52,11 @@ DevJourney.iml
 /force-app/main/default/lwc/accountsComponent/__tests__/
 /.husky/
 /.vscode/
+/DevJourney/
```
@AmigoBongero Is it OK to have it here?

AmigoBongero replied:
It is IDEA'S folder, wasn't needed for the project, but I changed gitignore to file inside this folder.

IJWMG commented on DevJourney/force-app/main/default/classes/AccountController.cls:
```diff
@@ -0,0 +1,22 @@
+public with sharing class AccountController {
+    @AuraEnabled
+    public static List<Account> getMyAccounts(){
```
@AmigoBongero Space between ) and {

IJWMG commented on DevJourney/force-app/main/default/classes/AccountController.cls:
```diff
@@ -0,0 +1,22 @@
+public with sharing class AccountController {
+    @AuraEnabled
```
@AmigoBongero Add one empty line between first and last class lines and actual code ("public ... class ... {" and last "}")

IJWMG commented on DevJourney/force-app/main/default/classes/AccountController.cls:
```diff
@@ -0,0 +1,22 @@
+public with sharing class AccountController {
```
@AmigoBongero Generally we name an Apex controller after the LWC or module that the class controlls. So here it should be something like "AccountComponentController"

IJWMG commented on DevJourney/force-app/main/default/classes/AccountController.cls:
```diff
@@ -0,0 +1,22 @@
+    @AuraEnabled
+    public static List<Account> getMyAccounts(){
+        return [
+                SELECT Id,Name, Industry, Phone
```
@AmigoBongero Whitespace between Id and Name

IJWMG commented on DevJourney/force-app/main/default/classes/AccountController.cls:
```diff
@@ -0,0 +1,22 @@
+        return [
+                SELECT Id,Name, Industry, Phone
+                FROM Account
+                WHERE OwnerId = :UserInfo.getUserId()
```
@AmigoBongero For this kind of variables when you get one through a method call, it's better to use a variable. I would recomend using static final variable, as userId is never changed during the transaction.

IJWMG commented on DevJourney/force-app/main/default/classes/AccountController.cls:
```diff
@@ -0,0 +1,22 @@
+                SELECT Id,Name, Industry, Phone
+                FROM Account
+                WHERE OwnerId = :UserInfo.getUserId()
+                WITH SECURITY_ENFORCED
```
@AmigoBongero Can you please explain why did you used WITH SECURITY_ENFORCED here?

AmigoBongero replied:
I thought that this is the best practise, as it ensures that the query respects the user's field-level security and object-level security settings.

IJWMG commented on DevJourney/force-app/main/default/classes/AccountController.cls:
```diff
@@ -0,0 +1,22 @@
+                WHERE OwnerId = :UserInfo.getUserId()
+                WITH SECURITY_ENFORCED
+                ORDER BY Name
+                LIMIT 500];
```
@AmigoBongero IMHO if you start code block with new line like 
` if (condition) { \n first_line_of_code_in_block` (didn't find how to use multiline code here, so \n stands for "New line")
you should end it with new line as well:
`your_last_line_of_code \n }` 
For real example see screenshot from the comment above :)

IJWMG commented on DevJourney/force-app/main/default/classes/AccountController.cls:
```diff
@@ -0,0 +1,22 @@
+public with sharing class AccountController {
+    @AuraEnabled
+    public static List<Account> getMyAccounts(){
+        return [
```
@AmigoBongero Use tabs to separate SOQL keywords from field/object names in a way, all field/object names starts from the same horizontal position. Something like this: 
![image](https://github.com/user-attachments/assets/60a68b34-dd98-45d1-ac93-36017456431e)

IJWMG commented on DevJourney/force-app/main/default/classes/AccountController.cls:
```diff
@@ -0,0 +1,22 @@
+                WITH SECURITY_ENFORCED
+                ORDER BY Name
+                LIMIT 500];
+    }
```
@AmigoBongero Methods inside the class should be separated by an empty line

IJWMG commented on DevJourney/force-app/main/default/classes/AccountController.cls:
```diff
@@ -0,0 +1,22 @@
+                LIMIT 500];
+    }
+    @AuraEnabled
+    public static List<Account> getRecentlyViewedAccounts(){
```
@AmigoBongero Apply the same changes as fro previous method please

IJWMG commented on DevJourney/force-app/main/default/classes/AccountController.cls:
```diff
@@ -0,0 +1,22 @@
+                WHERE OwnerId = :UserInfo.getUserId()
+                WITH SECURITY_ENFORCED
+                ORDER BY Name
+                LIMIT 500];
```
@AmigoBongero Why did you choose to set the limit to 500 here?

AmigoBongero replied:
It's just approximate number. I heard that it is better to set the limits in such queries, as there can be too many records in the org.

IJWMG commented on DevJourney/force-app/main/default/classes/AccountController.cls:
```diff
@@ -0,0 +1,22 @@
+                WHERE LastViewedDate != NULL
+                WITH SECURITY_ENFORCED
+                ORDER BY LastViewedDate DESC
+                LIMIT 50];
```
@AmigoBongero Why is this limit is 10 times less than "My Accounts" one?

AmigoBongero replied:
I considered that there's no sense for user to see more that 50 last viewed account records.

IJWMG replied:
@AmigoBongero Let me tell you about it on Monday. And for now, lets make limits equal in both cases

IJWMG commented on DevJourney/force-app/main/default/classes/AccountController.cls:
```diff
@@ -0,0 +1,22 @@
+        return [
+                SELECT Id, Name, Industry, Phone
+                FROM Account
+                WHERE LastViewedDate != NULL
```
@AmigoBongero This one will totally work, but did you consider other variants? Why did you choose this one above others?

AmigoBongero replied:
I found variant to return the list of recentlyViewedDate object with type = "account" but using this method we can't get such account fields as "industry", "phone". So, I found in the salesforce documentation that for this case we can do it like this.

IJWMG commented on DevJourney/force-app/main/default/lwc/accountsComponent/accountsComponent.html:
```diff
@@ -4,8 +4,24 @@
-                    <lightning-tab label="Recently viewed accounts"></lightning-tab>
+                    <lightning-tab label="My Accounts">
+                        <lightning-datatable
+                                key-field="id"
```
@AmigoBongero Please use one "tab" to mark inner tags and attributes

AmigoBongero replied:
Don't see where's the problem...

IJWMG commented on DevJourney/force-app/main/default/lwc/accountsComponent/accountsComponent.js:
```diff
@@ -1,4 +1,37 @@
-
+import { LightningElement} from 'lwc';
+import getMyAccounts from "@salesforce/apex/AccountController.getMyAccounts";
+import getRecentlyViewedAccounts from "@salesforce/apex/AccountController.getRecentlyViewedAccounts";
```
@AmigoBongero Group your imports please(divide them into groups separeted by an empty lines). Also add an empty line bewteen your imports and actual code

IJWMG commented on DevJourney/force-app/main/default/lwc/accountsComponent/accountsComponent.js:
```diff
@@ -1,4 +1,37 @@
+import { LightningElement} from 'lwc';
+import getMyAccounts from "@salesforce/apex/AccountController.getMyAccounts";
+import getRecentlyViewedAccounts from "@salesforce/apex/AccountController.getRecentlyViewedAccounts";
 export default class AccountsComponent extends LightningElement {
```
@AmigoBongero Add empty lines after class deffinition and before last "}" of the class please

IJWMG commented on DevJourney/force-app/main/default/lwc/accountsComponent/accountsComponent.js:
```diff
@@ -1,4 +1,37 @@
+import getMyAccounts from "@salesforce/apex/AccountController.getMyAccounts";
+import getRecentlyViewedAccounts from "@salesforce/apex/AccountController.getRecentlyViewedAccounts";
 export default class AccountsComponent extends LightningElement {
+    columns = [
```
@AmigoBongero If there is unchangeable variable (for example columns deffinition 😄) we can use constant variable for it.

IJWMG commented on DevJourney/force-app/main/default/lwc/accountsComponent/accountsComponent.js:
```diff
@@ -1,4 +1,37 @@
+    connectedCallback() {
+        this.fetchMyAccounts();
+        this.fetchRecentlyViewedAccounts();
+    }
```
@AmigoBongero So we call this once. Don't you think @wire adaptor fits here?

AmigoBongero replied:
I've tried using @wire adaptor, but there have been some problems with cacheable (as it has to be true when I use wire). But imperative apex call relosolved it, and now (I've tested it) when, for instance, I'm trying to create new Account and go back to accountsComponent the data dynamically changes. I considered that this is the right method for this case. That's why I decided not to use @wire here.

IJWMG commented on DevJourney/force-app/main/default/lwc/accountsComponent/accountsComponent.js:
```diff
@@ -1,4 +1,37 @@
+                this.myAccountsData = result;
+            })
+            .catch(error => {
+                console.error(error);
```
@AmigoBongero Let's use toast events to let user know something went wrong if it does

IJWMG commented on DevJourney/force-app/main/default/lwc/accountsComponent/accountsComponent.js:
```diff
@@ -1,4 +1,37 @@
+        this.fetchRecentlyViewedAccounts();
+    }
+
+    fetchMyAccounts() {
```
@AmigoBongero Both of the methods call apex method, which is not exactly a fetch by JS deffinition. Lets use more precise naming for them.

IJWMG commented on DevJourney/force-app/main/default/lwc/accountsComponent/accountsComponent.js:
```diff
@@ -1,4 +1,37 @@
+            });
+    }
+
+    fetchRecentlyViewedAccounts() {
```
@AmigoBongero Change in acccordance with previous comments please

IJWMG reviewed (commented):
@AmigoBongero Add new components to Permission set. Also please change the permission set to have LWC tab. And let's cover all the apex code with unit tests :)

AmigoBongero committed:
# Resolved all the corrections

- Fix: made corrections in accountsComponent
- Fix: made corrections in AccountsComponentController
- Fix: changed the path to the file which must be ignored in gitignore file
- Fix: made corrections to permset and added all the relevant metadata from the org
- Dev: wrote AccountComponentController Test

AmigoBongero reviewed (commented):
#Made all the corrections

IJWMG commented on DevJourney/force-app/main/default/applications/standard__LightningSales.app-meta.xml:
```diff
@@ -24,6 +24,10 @@
+    <tabs>standard-Case</tabs>
+    <tabs>standard-Forecasting3</tabs>
+    <tabs>Case_Party__c</tabs>
+    <tabs>Expense__c</tabs>
```
@AmigoBongero Fills too early to add this here)

IJWMG commented on DevJourney/force-app/main/default/lwc/accountsComponent/accountsComponent.js:
```diff
@@ -1,4 +1,56 @@
+import getMyAccounts from "@salesforce/apex/AccountsComponentController.getMyAccounts";
+import getRecentlyViewedAccounts from "@salesforce/apex/AccountsComponentController.getRecentlyViewedAccounts";
+
+const accountColumns = [
```
@AmigoBongero Consider constants naming please

IJWMG commented on DevJourney/force-app/main/default/lwc/accountsComponent/accountsComponent.js:
```diff
@@ -1,4 +1,56 @@
+            .catch(error => {
+                const showError = new ShowToastEvent({
+                    title: 'Error',
+                    message: 'Error loading accounts.',
```
@AmigoBongero I think it should either be 'Error occurred while loading accounts' or 'Accounts loading error'. And it's definitely should be title. Les't use real error message for toast message here

IJWMG commented on DevJourney/force-app/main/default/objects/Case_Party__c/listViews/All.listView-meta.xml:
@AmigoBongero Do you need it here?

IJWMG commented on DevJourney/force-app/main/default/objects/Expense__c/listViews/All.listView-meta.xml:
@AmigoBongero And this one too?)

IJWMG commented on DevJourney/force-app/main/default/classes/AccountsComponentControllerTest.cls:
```diff
@@ -0,0 +1,33 @@
+        Account testAccount = new Account(Name = 'Test Account', Industry = 'Technology', Phone = '123-456-7890');
+        insert testAccount;
+
+        System.runAs(new User(Id = UserInfo.getUserId())) {
```
@AmigoBongero There is no need to use System.runAs here

IJWMG commented on DevJourney/force-app/main/default/classes/AccountsComponentControllerTest.cls:
```diff
@@ -0,0 +1,33 @@
+            List<Account> myAccounts = AccountsComponentController.getMyAccounts();
+            Test.stopTest();
+
+            System.assertEquals(1, myAccounts.size(), 'Should have 1 account');
```
@AmigoBongero In general, it's better to test your code on multiple records, just to make sure it's bulkified

IJWMG replied:
@AmigoBongero Aslo, consider usin Assert class methods for test assertion

IJWMG commented on DevJourney/force-app/main/default/classes/AccountsComponentControllerTest.cls:
```diff
@@ -0,0 +1,33 @@
+
+    @IsTest
+    static void testGetMyAccounts() {
+        Account testAccount = new Account(Name = 'Test Account', Industry = 'Technology', Phone = '123-456-7890');
```
@AmigoBongero You don't have to fill in fields that you never using in the test

IJWMG commented on DevJourney/force-app/main/default/classes/AccountsComponentControllerTest.cls:
@AmigoBongero Please apply test comments structure to the test to improve readability. We are using this (\n stands for new line)
`// Prepare test data \n data_preparation_code \n \n // Execute \n test_execution \n \n // Assert results \n assert_code`

AmigoBongero committed:
# Made all the corrections

AmigoBongero reviewed (commented):
#Made all the previous corrections

IJWMG commented on DevJourney/force-app/main/default/classes/AccountsComponentControllerTest.cls:
```diff
@@ -0,0 +1,46 @@
+private class AccountsComponentControllerTest {
+
+    @IsTest
+    static void testGetMyAccounts() {
```
@AmigoBongero Name test methods like this: 'yourMethodNameTest' for positive tests and 'yourMethodNameNegativeTest' for negative tests.

IJWMG commented on DevJourney/force-app/main/default/classes/AccountsComponentControllerTest.cls:
```diff
@@ -0,0 +1,46 @@
+    static void testGetMyAccounts() {
+        // Prepare test data
+        List<Account> testAccounts = new List<Account>();
+        for (Integer i = 0; i < 5; i++) {
```
@AmigoBongero Use TestDataFactory for this (aslo use setup method) connect with @Asha2502 to sync TestDataFactories

IJWMG commented on DevJourney/force-app/main/default/classes/AccountsComponentControllerTest.cls:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    @IsTest
+    static void testGetRecentlyViewedAccounts() {
```
@AmigoBongero Naming

IJWMG commented on DevJourney/force-app/main/default/classes/AccountsComponentControllerTest.cls:
```diff
@@ -0,0 +1,46 @@
+        List<Account> testAccounts = new List<Account>();
+        for (Integer i = 0; i < 5; i++) {
+            Account testAccount = new Account(
+                    Name = 'Test Account' + i
```
@AmigoBongero Remind me to show you whats the metter with tabs here

IJWMG commented on DevJourney/force-app/main/default/lwc/accountsComponent/accountsComponent.js:
```diff
@@ -1,4 +1,56 @@
-import { LightningElement } from 'lwc';
+import {LightningElement} from 'lwc';
```
@AmigoBongero Nope, bring back the whitespaces :)

IJWMG commented on DevJourney/force-app/main/default/lwc/accountsComponent/accountsComponent.js:
```diff
@@ -1,4 +1,56 @@
 
 export default class AccountsComponent extends LightningElement {
+
+    accountColumns = ACCOUNT_COLUMNS;
```
@AmigoBongero Use getter please (in accordance with code guidelines)

IJWMG commented on DevJourney/force-app/main/default/objects/Case_Party__c/listViews/All.listView-meta.xml:
@AmigoBongero Same question :)

AmigoBongero replied:
Thought I removed them...

IJWMG commented on DevJourney/force-app/main/default/objects/Expense__c/listViews/All.listView-meta.xml:
@AmigoBongero And one more :)

AmigoBongero committed:
#Made corrections

- format: formatted accountsComponent.html, AccountsComponentController.cls
- dev: created TestSetup + Data Factory in AccountsComponentControllerTest.cls
- removed All.listView-meta files

AmigoBongero reviewed (commented):
- format: formatted accountsComponent.html, AccountsComponentController.cls
- dev: created TestSetup + Data Factory in AccountsComponentControllerTest.cls
- removed All.listView-meta files

IJWMG commented on DevJourney/force-app/main/default/classes/AccountsComponentControllerTest.cls:
```diff
@@ -0,0 +1,31 @@
+private class AccountsComponentControllerTest {
+
+    @TestSetup
+    static void loadTestData() {
```
@AmigoBongero Let's name it 'setup'

IJWMG commented on DevJourney/force-app/main/default/classes/AccountsComponentControllerTest.cls:
```diff
@@ -0,0 +1,31 @@
+        Test.stopTest();
+
+        // Assert results
+        Assert.areEqual(0, recentlyViewedAccounts.size(), 'Should have 0 accounts');
```
@AmigoBongero You better find a way to test this properly :) (hint: SOQL clause)

IJWMG commented on DevJourney/force-app/main/default/lwc/accountsComponent/accountsComponent.js:
```diff
@@ -1,4 +1,59 @@
-import { LightningElement } from 'lwc';
+
+import {LightningElement} from 'lwc';
```
@AmigoBongero Not an empty line, whitespaces 😄  (If you don't get it, check previous version in your own PR :))

IJWMG commented on DevJourney/force-app/main/default/lwc/accountsComponent/accountsComponent.js:
```diff
@@ -1,4 +1,59 @@
 
 export default class AccountsComponent extends LightningElement {
+
+    get accountColumns() {
```
@AmigoBongero Please name getters with 'Getter' posrfix.

AmigoBongero committed:
#Made corrections

AmigoBongero reviewed (commented):
- Renamed getter with postfix "Getter"
- Renamed TestSetup 
- Wrote correct assertion for getRecentlyViewedAccountsTest
- whitespaces...

AmigoBongero committed:
#Correctrions
